### PR TITLE
Added instawp integration

### PR DESCRIPTION
**Pull request by h-a-r-d-i-k**

### CODE REVIEW

File: includes/Extension/SearchReplace/class-replace.php [hunk 1]
```diff
@@ -16,11 +16,14 @@ class Replace {
...
 		add_action( 'string_locator_search_results_tablenav_controls', array( $this, 'output_replace_form' ) );
+		add_action( 'string_locator_instawp_tablenav_controls', array( $this, 'add_instawp_stage_button' ) );
 
 		add_action( 'admin_enqueue_scripts', array( $this, 'maybe_enqueue_assets' ) );
 
-		add_action( 'string_locator_search_templates', array( $this, 'add_replace_response_template' ) );
-	}
+		add_action( 'string_locator_search_templates', array( $this, 'add_replace_response_template' ) ); 
+		
+		add_action('wp_ajax_install_activate_plugin', array($this, 'install_activate_plugin_callback'));
```
Comment (Clorith) [inline includes/Extension/SearchReplace/class-replace.php]: This section need some spacing to match the WordPress coding standards

```suggestion
		add_action( 'wp_ajax_install_activate_plugin', array( $this, 'install_activate_plugin_callback' ) );
```

File: includes/Extension/SearchReplace/class-replace.php [hunk 2]
```diff
@@ -71,6 +74,14 @@ public function maybe_enqueue_assets( $hook ) {
...
 			)
 		);
+
+		/**
+		 * Instawp installation event handle script 
+		 * */
+		wp_enqueue_script( 'string-locator-instawp', trailingslashit( STRING_LOCATOR_PLUGIN_URL ) . 'build/string-locator-instawp.js', array( 'jquery', 'updates' ), false, 'all' );
+		wp_localize_script( 'string-locator-instawp', 'customScriptData', array(
+			'ajaxUrl' => admin_url( 'admin-ajax.php' ),
+		) );
```
Comment (Clorith) [inline includes/Extension/SearchReplace/class-replace.php]: This additional JavaScript object is never utilized, and could be removed, as the related code uses the `ajaxurl` variable added by WordPress core within all wp-admin pages.

File: includes/Extension/SearchReplace/class-replace.php [hunk 3]
```diff
@@ -85,12 +96,163 @@ public function add_replace_button() {
...
+			if ( ! empty( $classes ) ) {
+				$button['classes'] = array_merge( $button['classes'], $classes );
+			}
+
+			$button['classes'] = implode( ' ', $button['classes'] );
+
+			?>
+			<span class="plugin-card-<?php echo esc_attr( $plugin_slug ); ?>" style="float: right; margin-top: 7px;">
+				<?php echo !empty($button["logo-img"]) ? $button["logo-img"] : '';?>
+				<a href="<?php echo esc_url( $button['url'] ); ?>" class="<?php echo esc_attr( $button['classes'] ); ?>" data-originaltext="<?php echo esc_attr( $button['message'] ); ?>" data-name="<?php echo esc_attr( $plugin_name ); ?>" data-slug="<?php echo esc_attr( $plugin_slug ); ?>" aria-label="<?php echo esc_attr( $button['message'] ); ?>" <?php echo !empty($button["target"]) ? $button["target"] : '';?>>
```
Comment (Clorith) [inline includes/Extension/SearchReplace/class-replace.php]: The use of links for interactive buttons is not good for accessibility. These should be replaced with appropriate buttons and button styling instead. (this references the JavaScript later down the PR which grabs the links and treats them as buttons).

File: includes/Extension/SearchReplace/class-replace.php [hunk 4]
```diff
@@ -85,12 +96,163 @@ public function add_replace_button() {
...
+			<?php
+		}
+
+	}
+
+	/**
+	 * Handle Ajax call to activate the function
+	 *
+	 */
+	public function install_activate_plugin_callback()
```
Comment (Clorith) [inline includes/Extension/SearchReplace/class-replace.php:190]: This function handles an admin-ajax callback, but has no nonce verification or user authentication checks.

It also accepts a generic plugin slug as a post value, which may open up the plugin to allow any actor to install any plugin on a site with this plugin active.

This should be replaced with a REST endpoint like the other interactive elements, and also have the `install_plugins` capability check as part of that endpoints `permission_callback` check.

File: includes/Extension/SearchReplace/views/replace-form.php [hunk 5]
```diff
@@ -52,4 +52,7 @@
 			</button>
 		</p>
 	</form>
+	<?php if (!self::is_plugin_installed( $instawp_plugin )): ?>
+		<p class="notice notice-error update-nag">It is always a good idea to create a staging site before making changes <span id="notice-btn-toggle-replace-controls" class="button-link">Install Staging Plugin</span></p>
```
Comment (Clorith) [inline includes/Extension/SearchReplace/views/replace-form.php]: This text string is not wrapped in a translation function.

File: src/javascript/string-locator-instawp.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
+( function ( wp, $ ) {
```
Comment (Clorith) [inline src/javascript/string-locator-instawp.js:1]: The JavaScript code here adds a reliance on jQuery, but the other pieces of JavaScript uses vanilla JS to reduce dependencies.

File: includes/Extension/SearchReplace/views/replace-form.php (hunk 5 above)
Comment (Clorith) [inline includes/Extension/SearchReplace/views/replace-form.php]: An inline `span` element is treated as a button, which should instead be an actual button.

File: includes/Extension/SearchReplace/class-replace.php [hunk 7]
```diff
@@ -71,6 +74,14 @@ public function maybe_enqueue_assets( $hook ) {
 				),
 			)
 		);
+
+		/**
+		 * Instawp installation event handle script 
+		 * */
+		wp_enqueue_script( 'string-locator-instawp', trailingslashit( STRING_LOCATOR_PLUGIN_URL ) . 'build/string-locator-instawp.js', array( 'jquery', 'updates' ), false, 'all' );
```
Comment (Clorith) [inline includes/Extension/SearchReplace/class-replace.php]: A version string is missing to help avoid cached content when new releases of the plugin are deployed.

The final argument has the `all` keyword, which is not a valid entry,. this is a boolean entry to determine if the script should be enqueued in the footer of the site or not.

File: includes/Extension/SearchReplace/class-replace.php [hunk 8]
```diff
@@ -85,12 +96,163 @@ public function add_replace_button() {
...
+					'url'     => $instawp_connect,
+					'classes' => array( 'string-locator-instawp-button', 'disabled' ),
+				);
+
+				if ( '' !== $activated ) {
+					$button['message'] = esc_attr( $activated );
+
+					$button['target'] = 'target="_blank"';
+				}
+				$button['logo-img'] = '<span class="btn-logo-inline"><img src="'. plugins_url( 'views/assets/instawp-logo.svg', STRING_LOCATOR_PLUGIN_FILE ).'" alt="InstaWP logo"></span>';
```
Comment (Clorith) [inline includes/Extension/SearchReplace/class-replace.php]: The image source is not escaping the URL here.

File: includes/Extension/SearchReplace/class-replace.php [hunk 9]
```diff
@@ -85,12 +96,163 @@ public function add_replace_button() {
...
+
+				// The plugin exists but isn't activated yet.
+				$button = array(
+					'message' => esc_attr__( 'Create a Staging Site', 'string-locator' ),
+					// 'url'     => $url,
+					'url'     => "javascript:void 0;",
+					'classes' => array( 'instawp-activate-now' ),
+				);
+
+				$button['logo-img'] = '<span class="btn-logo-inline"><img src="'. plugins_url( 'views/assets/instawp-logo.svg', STRING_LOCATOR_PLUGIN_FILE ).'" alt="InstaWP logo"></span>';
```
Comment (Clorith) [inline includes/Extension/SearchReplace/class-replace.php]: The image source is not escaping the URL here.

File: src/javascript/string-locator-instawp.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+				if (response.success) {
+					$button.text(response.data.anchor_text);
+					$button.attr('href', response.data.href);
+					$button.attr('data-originaltext', response.data.anchor_text);
+					$button.attr('aria-label', response.data.anchor_text);
+					$button.attr('target', '_blank');
+					$button.removeClass('instawp-activate-now');
+					$button.addClass('string-locator-instawp-button disabled');
+				} else {
+					console.log('Error installing and activating the plugin:', response.data);
```
Comment (Clorith) [inline src/javascript/string-locator-instawp.js]: The error handling here requires a developer console to be open, which the target audience is unlikely to have.

File: src/javascript/string-locator-instawp.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+								$button.attr('aria-label', response.data.anchor_text);
+								$button.attr('target', '_blank');
+								$button.removeClass('sl-instawp-install-now install-now install-instawp-connect updating-message');
+								$button.addClass('string-locator-instawp-button disabled');
+								
+								$(".update-nag").remove();
+
+							}, 3000);
+						} else {
+							console.log('Error installing and activating the plugin:', response.data);
```
Comment (Clorith) [inline src/javascript/string-locator-instawp.js]: The error handling here requires a developer console to be open, which the target audience is unlikely to have.

File: src/javascript/string-locator-instawp.js [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+						}
+					});
+
+				},
+				error: function(error) {
+					console.log('Error installing the plugin:', error);
```
Comment (Clorith) [inline src/javascript/string-locator-instawp.js]: The error handling here requires a developer console to be open, which the target audience is unlikely to have.

File: views/search.php [hunk 13]
```diff
@@ -140,14 +140,27 @@
...
 		<div class="tablenav top">
-			<?php
-			/**
-			 * An action to output controls in the tablenav region, which only
-			 * become visible when there are search results available.
-			 */
-			do_action( 'string_locator_search_results_tablenav_controls' );
-			?>
-			<br class="clear" />
+			<div style="display: flex;">
```
Comment (Clorith) [inline views/search.php]: The flex style here makes the replacement controls be limited in width and looks "wrong" when expanding the replacement panel.

File: includes/Extension/SearchReplace/class-replace.php [hunk 14]
```diff
@@ -85,12 +96,163 @@ public function add_replace_button() {
...
+
+		$instawp_link = menu_page_url( 'instawp', false );
+		if(empty($instawp_link)) {
+			$instawp_link = admin_url('tools.php?page=instawp');
+		}
+
+		$response = array(
+			'message' => 'Plugin installed and activated successfully.',
+			'href' => $instawp_link,
+			'anchor_text' => 'Go to InstaWP →'
```
Comment (Clorith) [inline includes/Extension/SearchReplace/class-replace.php]: These strings are not translatable.

File: includes/Extension/SearchReplace/class-replace.php [hunk 15]
```diff
@@ -85,12 +96,163 @@ public function add_replace_button() {
...
+	 *
+	 */
+	public function install_activate_plugin_callback()
+	{
+		$pluginSlug = isset($_POST['plugin']) ? $_POST['plugin'] : '';
+
+		$pluginSlugFile = $pluginSlug.'/'.$pluginSlug.'.php';
+
+		if (empty($pluginSlug)) {
+			wp_send_json_error('Invalid plugin slug.');
```
Comment (Clorith) [inline includes/Extension/SearchReplace/class-replace.php]: This string is not translatable.

**Review by Clorith — CHANGES_REQUESTED** [review]

File: includes/Extension/SearchReplace/class-replace.php (hunk 1 above)
Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/class-replace.php]: Resolved In latest Commit.

File: includes/Extension/SearchReplace/class-replace.php (hunk 2 above)
Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/class-replace.php]: Resolved In latest Commit.

File: includes/Extension/SearchReplace/class-replace.php (hunk 3 above)
Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/class-replace.php]: Resolved In latest Commit.

File: includes/Extension/SearchReplace/views/replace-form.php (hunk 5 above)
Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/views/replace-form.php]: Resolved in latest commit.

Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/views/replace-form.php]: Reolved in latest commit.

File: includes/Extension/SearchReplace/class-replace.php (hunk 7 above)
Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/class-replace.php]: Reolved in latest commit.

File: includes/Extension/SearchReplace/class-replace.php (hunk 8 above)
Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/class-replace.php]: Reolved in latest commit.

File: includes/Extension/SearchReplace/class-replace.php (hunk 9 above)
Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/class-replace.php]: Resolved In latest commit.

File: src/javascript/string-locator-instawp.js (hunk 11 above)
Comment (h-a-r-d-i-k) [reply src/javascript/string-locator-instawp.js]: Resolved in latest commit.

File: src/javascript/string-locator-instawp.js (hunk 10 above)
Comment (h-a-r-d-i-k) [reply src/javascript/string-locator-instawp.js]: Resolved in latest commit.

File: src/javascript/string-locator-instawp.js (hunk 12 above)
Comment (h-a-r-d-i-k) [reply src/javascript/string-locator-instawp.js]: Resolved in latest commit.

File: includes/Extension/SearchReplace/class-replace.php (hunk 14 above)
Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/class-replace.php]: Resolved in latest commit.

File: includes/Extension/SearchReplace/class-replace.php (hunk 15 above)
Comment (h-a-r-d-i-k) [reply includes/Extension/SearchReplace/class-replace.php]: Resolved in latest commit.

**Pull request merged**